opnode/rollup/sync/start.go

```diff
@@ -62,96 +60,139 @@ var TooDeepReorgErr = errors.New("reorg is too deep")
 
 const MaxReorgDepth = 500
 
-// isCanonical returns true if the supplied block ID is canonical in the L1 chain.
-// It will suppress ethereum.NotFound errors
-func isCanonical(ctx context.Context, l1 L1Chain, block eth.BlockID) (bool, error) {
-	canonical, err := l1.L1BlockRefByNumber(ctx, block.Number)
-	if err != nil && !errors.Is(err, ethereum.NotFound) {
-		return false, err
-	} else if err != nil {
-		return false, nil
+// isCanonical returns the following values:
+// - `aheadOrCanonical: true if the supplied block is ahead of the known head of the L1 chain,
+//    or canonical in the L1 chain.
+// - `canonical`: true if the block is canonical in the L1 chain.
+func isAheadOrCanonical(ctx context.Context, l1 L1Chain, block eth.BlockID) (aheadOrCanonical bool, canonical bool, err error) {
+	if l1Head, err := l1.L1HeadBlockRef(ctx); err != nil {
```

Comment:
I don't love that we call `L1HeadBlockRef` on every iteration. I think it could probably be called once, but it may have an interaction with `BlockRefByNumber`

What could happen is that the block could be ahead of the head at one point, and then at a later point turn out to be canonical - as such I think it needs to do this. Still a bit weird to have the L1 head maybe updating underneath us in this look.

Reply:
I briefly considered using the "block not found" error as a signal that the block was ahead, but it feels a bit dirty.

I think we could get away with doing it once — both ahead blocks & canonical blocks can be the unsafe head. This might lead us to rollback too far for the safe head, but I think that's generally a safe thing to do (lol).

(Note that the block being ahead then canonical could happen in the current function too! Could be good to include a comment that explains we considered this pitfall.)

I'm not sure we should change it though. I think `l1.L1HeadBlockRef(ctx)` should always return a cached value (and if not, we should fix that), so the overhead should be neglectible.